### ISSUE TITLE
Unexpected performance  of dropout

### ISSUE THREAD

**Issue by ang868** [pytorch/xla#3868]
hi, recently I've been doing some torch xla performance analysis on gpu.  I came across with a problem that the default random algorithm with poor performance. 

I use bert base written by huggingface transformers as my test algorithm and all of the following tests were done on nvidia 16G V100.

The througput results training with torch xla are as follows:

- bert base without dropout: 150.38 samples/s
- bert base with dropout:       106.67 samples/s

There are 29.1% performance drop. The magnitude of the performance degradation does not meet expectations. I tested bert base running with pytorch native:

- bert base without dropout: 99.62 samples/s
- bert base  with dropout:      91.23 samples/s

There are only 8.4% performance drop.

I found there are three options to select different  random algorithm in torch xla: default, philox and three_fry. So I tested the performance impact of the three different algorithms:

- ThreeFry generator(torch xla):  129.87 samples/s
- Philox generator (torch xla):      102.56 samples/s
- Default generator (torch xla):    106.67 samples/s

I think the philox generator will eventually be selected if you choose default generator option. So they're basically the same performance. 

So my questions are:

1. As described in the tensorflow document(https://www.tensorflow.org/guide/random_numbers#xla_devices), ThreeFry is fast on TPU but slow on CPU/GPU compared to Philox. Is it as expected that three fry generator has better throughput performance than philox? 
2. Although the performance of three fry is better than philox, it is still worse than expected. Is there any way to further optimize this performance?

**Comment by JackCaoG** [comment]
Thanks for raising this issue, I will try to find someone from xla:gpu team to answer this question.

**Comment by ymwangg** [comment]
@ang868 I saw very similar patterns before. XLA implements these RNGs using other XLA ops (e.g. [here](https://github.com/tensorflow/tensorflow/blob/490246643af1043ead0f5584158c062a63004012/tensorflow/compiler/xla/client/lib/prng.cc#L46-L49)) and the generated cuda kernel may not be as performant as those pytorch handwritten kernels that utilizes curand library. I'm not sure why XLA does not have a curand version of RNG but my guess is XLA wants a stateless RNG. One optimization I came up with is to replace the torch_xla RNG with the pytorch version using XLA custom_call, but the main issue I found is using XLA custom_call may reduce fusion opportunities and thus deliver limited speedup while increases memory usage.

BTW the performance of three_fry looks very reasonable and I think we may consider making it default on GPU.

**Comment by ang868** [comment]
@ymwangg thanks for your reply. yeah, we use three_fry generator as default for now. Hope that one day we could use the curand RNG wihout memory problems. ^_^

**Comment by ang868** [comment]
set three_fry as the default RNG for GPU.

**Comment by hgt312** [comment]
In addition to performance, it seems to have a significant impact on results.

I try to use torch_xla with DeepSpeed (with default generator), I get always inf loss with ZeRO-1 and dozens of iteration inf loss with ZeRO-3. If I switch to threeFry generator, it works well
